src/transformers/models/whisper/generation_whisper.py

```diff
@@ -884,39 +905,82 @@ def _prepare_segments(prompt_ids, batch_size, generation_config):
+        self, seek_outputs, decoder_input_ids, return_token_timestamps, generation_config, is_shortform
+    ):
         # remove all previously passed decoder input ids
         if isinstance(seek_outputs, torch.Tensor):
-            seek_outputs = seek_outputs[:, decoder_input_ids.shape[-1] :]
-            return seek_outputs, seek_outputs
+            if not is_shortform:
+                seek_outputs = seek_outputs[:, decoder_input_ids.shape[-1] :]
+                return seek_outputs, seek_outputs
+            else:
+                return seek_outputs, seek_outputs
 
         if return_token_timestamps and hasattr(generation_config, "alignment_heads"):
             num_frames = getattr(generation_config, "num_frames", None)
             seek_outputs["token_timestamps"] = self._extract_token_timestamps(
                 seek_outputs, generation_config.alignment_heads, num_frames=num_frames
             )
-            seek_outputs["token_timestamps"] = seek_outputs["token_timestamps"][:, decoder_input_ids.shape[-1] :]
+            if not is_shortform:
+                seek_outputs["token_timestamps"] = seek_outputs["token_timestamps"][:, decoder_input_ids.shape[-1] :]
 
-        seek_outputs["sequences"] = seek_outputs["sequences"][:, decoder_input_ids.shape[-1] :]
+        if not is_shortform:
+            seek_outputs["sequences"] = seek_outputs["sequences"][:, decoder_input_ids.shape[-1] :]
 
-        def split_by_batch_index(values, key, batch_idx):
-            if key == "scores":
+        def split_by_batch_index(values, key, batch_idx, is_shortform):
+            if key in ["scores", "encoder_attentions", "encoder_hidden_states", "logits"]:
                 return [v[batch_idx].cpu() for v in values]
-            elif key == "past_key_values":
-                # we don't save `past_key_values` as this is too costly
-                return None
-            elif isinstance(values[batch_idx], tuple) and torch.is_tensor(values[batch_idx][0]):
+            if key in ["decoder_attentions", "decoder_hidden_states", "cross_attentions"]:
                 return tuple(tuple(w[batch_idx][None].cpu() for w in v) for v in values)
+            elif key == "past_key_values":
+                if not is_shortform:
+                    # we don't save `past_key_values` as this is too costly for longform
+                    return None
+                else:
+                    return tuple(tuple(w[batch_idx][None].cpu() for w in v) for v in values)
+
             return values[batch_idx].cpu()
 
         sequence_tokens = seek_outputs["sequences"]
+
         seek_outputs = [
-            {k: split_by_batch_index(v, k, i) for k, v in seek_outputs.items()}
+            {k: split_by_batch_index(v, k, i, is_shortform) for k, v in seek_outputs.items()}
             for i in range(sequence_tokens.shape[0])
         ]
 
         return sequence_tokens, seek_outputs
 
+    def _stack_split_outputs(self, seek_outputs, model_output_type, device):
```

Comment:
Only nit with this method is that we stack everything on CPU, and then move over to GPU again

=> could you explore quickly whether there's a way to create the stacked tensors directly on device?

Otherwise, the CPU <-> GPU transfer is going to be quite slow

Reply:
is this resolved?

Reply:
also would be nice to documment 1 ligne about why we need this method